pkg/koordlet/resmanager/memory_evict.go

```diff
@@ -0,0 +1,168 @@
+		return
+	}
+
+	memoryCapacity := node.Status.Capacity.Memory().Value()
+	if memoryCapacity <= 0 {
+		klog.Warningf("skip memory evict, memory capacity(%v) should greater than 0", memoryCapacity)
+		return
+	}
+
+	nodeMemoryUsage := nodeMetric.MemoryUsed.MemoryWithoutCache.Value() * 100 / memoryCapacity
+	if nodeMemoryUsage < *thresholdPercent {
+		klog.Warningf("skip memory evict, node memory usage(%v) is below threshold(%v)", nodeMemoryUsage, thresholdConfig)
+		return
+	}
+
+	klog.Infof("node(%v) MemoryUsage(%v): %.2f, evictThresholdUsage: %.2f",
+		m.resManager.nodeName,
+		nodeMetric.MemoryUsed.MemoryWithoutCache.Value(),
+		float64(nodeMemoryUsage)/100,
+		float64(*thresholdPercent)/100,
+	)
+
+	lowPercent := *thresholdPercent - memoryReleaseBufferPercent
+	memoryNeedRelease := memoryCapacity * (nodeMemoryUsage - lowPercent) / 100
+	m.killAndEvictBEPods(node, podMetrics, memoryNeedRelease)
+}
+
+func (m *MemoryEvictor) killAndEvictBEPods(node *corev1.Node, podMetrics []*metriccache.PodResourceMetric, memoryNeedRelease int64) {
+	bePodInfos := m.getSortedPodInfos(podMetrics)
+	message := fmt.Sprintf("killAndEvictBEPods for node(%v), need to release memory: %v", m.resManager.nodeName, memoryNeedRelease)
+	memoryReleased := int64(0)
+
+	var killedPods []*corev1.Pod
+	for _, bePod := range bePodInfos {
+		if memoryReleased < memoryNeedRelease {
+			killMsg := fmt.Sprintf("%v, kill pod: %v", message, bePod.pod.Name)
+			killContainers(bePod.pod, killMsg)
+			killedPods = append(killedPods, bePod.pod)
+			if bePod.podMetric != nil {
+				memoryReleased += bePod.podMetric.MemoryUsed.MemoryWithoutCache.Value()
+			}
+		}
+	}
+
+	for _, pod := range killedPods {
+		m.resManager.evictPod(pod, node, evictPodByNodeMemoryUsage, message)
+	}
+
+	m.lastEvictTime = time.Now()
+	klog.Infof("killAndEvictBEPods completed, memoryNeedRelease(%v) memoryReleased(%v)", memoryNeedRelease, memoryNeedRelease)
+}
+
+func (m *MemoryEvictor) getSortedPodInfos(podMetrics []*metriccache.PodResourceMetric) []*podInfo {
+	podMetricMap := make(map[string]*metriccache.PodResourceMetric, len(podMetrics))
+	for _, podMetric := range podMetrics {
+		podMetricMap[podMetric.PodUID] = podMetric
+	}
+
+	var bePodInfos []*podInfo
+	for _, podMeta := range m.resManager.statesInformer.GetAllPods() {
+		pod := podMeta.Pod
+		if extension.GetPodQoSClass(pod) == extension.QoSBE {
+			info := &podInfo{
+				pod:       pod,
+				podMetric: podMetricMap[string(pod.UID)],
+			}
+			bePodInfos = append(bePodInfos, info)
+		}
+	}
+
+	sort.Slice(bePodInfos, func(i, j int) bool {
+		if bePodInfos[i].pod.Spec.Priority == nil || bePodInfos[j].pod.Spec.Priority == nil ||
+			*bePodInfos[i].pod.Spec.Priority == *bePodInfos[j].pod.Spec.Priority {
+			return bePodInfos[i].podMetric.MemoryUsed.MemoryWithoutCache.Value() > bePodInfos[j].podMetric.MemoryUsed.MemoryWithoutCache.Value()
+		}
+		return *bePodInfos[i].pod.Spec.Priority < *bePodInfos[j].pod.Spec.Priority
```

Comment:
should consider koordinator priority

Reply:
If the priority is the same, resource specifications and running time need to be considered. Here is where the expansion strategy is required. Abstracting a sorted interface is convenient for expansion?

Reply:
leave a todo here